pkg/storage/chunk/client/baidubce/bos_storage_client.go

```diff
@@ -0,0 +1,171 @@
+//			 https://intl.cloud.baidu.com/doc/BOS/s/Ajwvysfpl-en
+const NoSuchKeyErr = "NoSuchKey"
+
+const DefaultEndpoint = bos.DEFAULT_SERVICE_DOMAIN
+
+var bosRequestDuration = instrument.NewHistogramCollector(prometheus.NewHistogramVec(prometheus.HistogramOpts{
+	Namespace: "loki",
+	Name:      "bos_request_duration_seconds",
+	Help:      "Time spent doing bos requests.",
+	Buckets:   prometheus.ExponentialBuckets(0.005, 4, 6),
+}, []string{"operation", "status_code"}))
+
+func init() {
+	bosRequestDuration.Register()
+}
+
+type BOSStorageConfig struct {
+	BucketName      string `yaml:"bucket_name"`
+	Endpoint        string `yaml:"endpoint"`
+	AccessKeyID     string `yaml:"access_key_id"`
+	SecretAccessKey string `yaml:"secret_access_key"`
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *BOSStorageConfig) RegisterFlags(f *flag.FlagSet) {
+	cfg.RegisterFlagsWithPrefix("", f)
+}
+
+// RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet
+func (cfg *BOSStorageConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.StringVar(&cfg.BucketName, prefix+"baidubce.bucket-name", "", "Name of BOS bucket.")
+	f.StringVar(&cfg.Endpoint, prefix+"baidubce.endpoint", DefaultEndpoint, "BOS endpoint to connect to.")
+	f.StringVar(&cfg.AccessKeyID, prefix+"baidubce.access-key-id", "", "Baidu BCE Access Key ID.")
+	f.StringVar(&cfg.SecretAccessKey, prefix+"baidubce.secret-access-key", "", "Baidu BCE Secret Access Key.")
+}
+
+type BOSObjectStorage struct {
+	cfg    *BOSStorageConfig
+	client *bos.Client
+}
+
+func NewBOSObjectStorage(cfg *BOSStorageConfig) (*BOSObjectStorage, error) {
+	clientConfig := bos.BosClientConfiguration{
+		Ak:               cfg.AccessKeyID,
+		Sk:               cfg.SecretAccessKey,
+		Endpoint:         cfg.Endpoint,
+		RedirectDisabled: false,
+	}
+	bosClient, err := bos.NewClientWithConfig(&clientConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &BOSObjectStorage{
+		cfg:    cfg,
+		client: bosClient,
+	}, nil
+}
+
+func (b *BOSObjectStorage) PutObject(ctx context.Context, objectKey string, object io.ReadSeeker) error {
+	return instrument.CollectedRequest(ctx, "BOS.PutObject", bosRequestDuration, instrument.ErrorCode, func(ctx context.Context) error {
+		body, err := bce.NewBodyFromSizedReader(object, -1)
+		if err != nil {
+			return err
+		}
+		_, err = b.client.BasicPutObject(b.cfg.BucketName, objectKey, body)
+		return err
+	})
+}
+
+func (b *BOSObjectStorage) GetObject(ctx context.Context, objectKey string) (io.ReadCloser, int64, error) {
+	var res *api.GetObjectResult
+	err := instrument.CollectedRequest(ctx, "BOS.GetObject", bosRequestDuration, instrument.ErrorCode, func(ctx context.Context) error {
+		var requestErr error
+		res, requestErr = b.client.BasicGetObject(b.cfg.BucketName, objectKey)
+		return requestErr
+	})
+	if err != nil {
+		return nil, 0, errors.Wrapf(err, "failed to get bos object [ %s ]", objectKey)
```

Comment:
ok, tks, This part of the code is not unit tested, it will just show up in the log of error reporting and will not have an impact. If it can pass CI, it is no problem.